x/bounty/keeper/msg_server.go

```diff
@@ -0,0 +1,424 @@
+}
+
+func (k msgServer) EditFinding(goCtx context.Context, msg *types.MsgEditFinding) (*types.MsgEditFindingResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	program, isExist := k.GetProgram(ctx, msg.ProgramId)
+	if !isExist {
+		return nil, types.ErrProgramNotExists
+	}
+	if program.Status != types.ProgramStatusActive {
+		return nil, types.ErrProgramNotActive
+	}
+
+	finding, found := k.GetFinding(ctx, msg.FindingId)
+	if !found {
+		return nil, types.ErrFindingNotExists
+	}
+	// check submitter
+	if finding.SubmitterAddress != msg.SubmitterAddress {
+		return nil, types.ErrFindingSubmitterInvalid
+	}
+	if len(msg.Title) > 0 {
+		finding.Title = msg.Title
+	}
+	if len(msg.FindingHash) > 0 {
+		finding.FindingHash = msg.FindingHash
+	}
+	if msg.SeverityLevel != types.Unspecified {
+		finding.SeverityLevel = msg.SeverityLevel
+	}
+	if len(msg.Detail) > 0 {
+		finding.Detail = msg.Detail
+	}
+
+	k.SetFinding(ctx, finding)
+
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			types.EventTypeEditFinding,
+			sdk.NewAttribute(types.AttributeKeyFindingID, finding.FindingId),
+			sdk.NewAttribute(types.AttributeKeyProgramID, finding.ProgramId),
+		),
+		sdk.NewEvent(
+			sdk.EventTypeMessage,
+			sdk.NewAttribute(sdk.AttributeKeyModule, types.AttributeValueCategory),
+			sdk.NewAttribute(sdk.AttributeKeySender, msg.SubmitterAddress),
+		),
+	})
+
+	return &types.MsgEditFindingResponse{}, nil
+}
+
+func (k msgServer) ConfirmFinding(goCtx context.Context, msg *types.MsgConfirmFinding) (*types.MsgConfirmFindingResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	finding, err := k.hostProcess(ctx, msg)
+	if err != nil {
+		return nil, err
+	}
+
+	finding.Status = types.FindingStatusConfirmed
+	k.SetFinding(ctx, finding)
+
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			types.EventTypeConfirmFinding,
+			sdk.NewAttribute(types.AttributeKeyFindingID, finding.FindingId),
+			sdk.NewAttribute(types.AttributeKeyProgramID, finding.ProgramId),
+		),
+		sdk.NewEvent(
+			sdk.EventTypeMessage,
+			sdk.NewAttribute(sdk.AttributeKeyModule, types.AttributeValueCategory),
+			sdk.NewAttribute(sdk.AttributeKeySender, msg.OperatorAddress),
+		),
+	})
+
+	return &types.MsgConfirmFindingResponse{}, nil
+}
+
+func (k msgServer) hostProcess(ctx sdk.Context, msg *types.MsgConfirmFinding) (types.Finding, error) {
+	var finding types.Finding
+	// get finding
+	finding, found := k.GetFinding(ctx, msg.FindingId)
+	if !found {
+		return finding, types.ErrFindingNotExists
+	}
+	// get program
+	program, isExist := k.GetProgram(ctx, finding.ProgramId)
+	if !isExist {
+		return finding, types.ErrProgramNotExists
+	}
+	if program.Status != types.ProgramStatusActive {
+		return finding, types.ErrProgramNotActive
+	}
+
+	// only host can update finding comment
+	if program.AdminAddress != msg.OperatorAddress {
+		return finding, types.ErrProgramCreatorInvalid
+	}
+
+	// fingerprint comparison
+	bz := k.cdc.MustMarshal(&finding)
```

Comment:
finding.Description and finding.ProofOfConcept should be excluded from the hash calculation.